# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/media/webrtc/jsapi/PeerConnectionImpl.cpp
# Commit: 99ea4f221834
# Full Hash: 99ea4f221834f706be3738a92183287b54d8f718
# Author: Byron Campen [:bwc] <docfaraday@gmail.com>
# Date: 2022-12-22 17:00:32
# Regressor Bug: 1800136
# File Overlap Count: 1
# Description:
#   Bug 1800136: Make RTCRtpSender::GetStats get ICE stats. r=ng
#   
#   Also fix bugs where we would get duplicate ICE stats. This mostly mattered to
#   WebrtcGlobalInformation since RTCStatsReportInternal does not automatically
#   remove duplicate ids like RTCStatsReport (RTCStatsReport is maplike).
# ==============================================================================

diff -r 5801060dd3d0 -r 99ea4f221834 dom/media/webrtc/jsapi/PeerConnectionImpl.cpp
--- a/dom/media/webrtc/jsapi/PeerConnectionImpl.cpp	Thu Dec 22 02:33:56 2022 +0000
+++ b/dom/media/webrtc/jsapi/PeerConnectionImpl.cpp	Thu Dec 22 02:33:57 2022 +0000
@@ -3198,6 +3198,13 @@
   DOMHighResTimeStamp now = mTimestampMaker.GetNow();
 
   nsTArray<dom::RTCCodecStats> codecStats = GetCodecStats(now);
+  std::set<std::string> transportIds;
+
+  if (!aSelector) {
+    // There might not be any senders/receivers if we're DataChannel only, so we
+    // don't handle the null selector case in the loop below.
+    transportIds.insert("");
+  }
 
   nsTArray<
       std::tuple<RTCRtpTransceiver*, RefPtr<RTCStatsPromise::AllPromiseType>>>
@@ -3208,19 +3215,24 @@
     if (!sendSelected && !recvSelected) {
       continue;
     }
+
+    if (aSelector) {
+      transportIds.insert(transceiver->GetTransportId());
+    }
+
     nsTArray<RefPtr<RTCStatsPromise>> rtpStreamPromises;
     // Get all rtp stream stats for the given selector. Then filter away any
     // codec stat not related to the selector, and assign codec ids to the
     // stream stats.
+    // Skips the ICE stats; we do our own queries based on |transportIds| to
+    // avoid duplicates
     if (sendSelected) {
       rtpStreamPromises.AppendElements(
-          transceiver->Sender()->GetStatsInternal());
+          transceiver->Sender()->GetStatsInternal(true));
     }
     if (recvSelected) {
-      // Right now, returns two promises; one for RTP/RTCP stats, and
-      // another for ICE stats.
       rtpStreamPromises.AppendElements(
-          transceiver->Receiver()->GetStatsInternal());
+          transceiver->Receiver()->GetStatsInternal(true));
     }
     transceiverStatsPromises.AppendElement(
         std::make_tuple(transceiver.get(),
@@ -3231,17 +3243,8 @@
   promises.AppendElement(RTCRtpTransceiver::ApplyCodecStats(
       std::move(codecStats), std::move(transceiverStatsPromises)));
 
-  // TODO(bug 1616937): We need to move this is RTCRtpSender, to make
-  // getStats on those objects work properly. It might be worth optimizing
-  // the null selector case, so we don't end up with bunches of copies of
-  // the same transport information in the final report.
-  if (aSelector) {
-    std::string transportId = GetTransportIdMatchingSendTrack(*aSelector);
-    if (!transportId.empty()) {
-      promises.AppendElement(mTransportHandler->GetIceStats(transportId, now));
-    }
-  } else {
-    promises.AppendElement(mTransportHandler->GetIceStats("", now));
+  for (const auto& transportId : transportIds) {
+    promises.AppendElement(mTransportHandler->GetIceStats(transportId, now));
   }
 
   promises.AppendElement(GetDataChannelStats(mDataConnection, now));